File: exercises/twelve-days/.meta/src/reference/java/TwelveDays.java
public class TwelveDays {
    private static String[] days = new String[]{"first", "second", "third", "fourth", "fifth", "sixth", "seventh",
        "eighth", "ninth", "tenth", "eleventh", "twelfth"};
    private static String[] gifts = new String[]{
        "a Partridge in a Pear Tree.",
        "two Turtle Doves, ",
        "three French Hens, ",
        "four Calling Birds, ",
        "five Gold Rings, ",
        "six Geese-a-Laying, ",
        "seven Swans-a-Swimming, ",
        "eight Maids-a-Milking, ",
        "nine Ladies Dancing, ",
        "ten Lords-a-Leaping, ",
        "eleven Pipers Piping, ",
        "twelve Drummers Drumming, "
    };

    public String verse(int verseNumber) {
        StringBuilder stringBuilder = new StringBuilder("On the " + days[verseNumber - 1]
                + " day of Christmas my true love gave to me: ");
        for (int i = verseNumber; i > 0; i--) {
            if (verseNumber != 1 && i == 1) {
                stringBuilder.append("and ");
            }
            stringBuilder.append(gifts[i - 1]);
        }
        stringBuilder.append("\n");
        return stringBuilder.toString();
    }

    public String verses(int verseNumberStart, int verseNumberEnd) {
        StringBuilder stringBuilder = new StringBuilder(verse(verseNumberStart));
        for (int i = verseNumberStart + 1; i <= verseNumberEnd; i++) {
            stringBuilder.append("\n");
            stringBuilder.append(verse(i));
        }
        return stringBuilder.toString();
    }

    public String sing() {
        return verses(1, 12);
    }
}

File: go/twelve-days/twelve_days.go
package twelve

import (
	"strings"
)

var items = []struct{
	day int
	dayOrdinal string
	quantifier string
	itemName string
}{
	{1, "first", "one","a Partridge in a Pear Tree"},
	{2, "second","two","Turtle Doves"},
	{3, "third","three","French Hens"},
	{4, "fourth","four","Calling Birds"},
	{5, "fifth","five","Gold Rings"},
	{6, "sixth","six","Geese-a-Laying"},
	{7, "seventh","seven","Swans-a-Swimming"},
	{8, "eighth","eight","Maids-a-Milking"},
	{9, "ninth","nine","Ladies Dancing"},
	{10, "tenth","ten","Lords-a-Leaping"},
	{11, "eleventh","eleven","Pipers Piping"},
	{12, "twelfth","twelve","Drummers Drumming"},
}

func Verse(n int) string {
	var s strings.Builder

	s.WriteString("On the " + items[n-1].dayOrdinal + " day of Christmas my true love gave to me: ")

	for i := n-1; i>=0; i-- {
		if i == 0 {
			if n != 1{
				s.WriteString( "and " + items[i].itemName)
			} else {
				s.WriteString( items[i].itemName)
			}

		} else {
			s.WriteString( items[i].quantifier + " " + items[i].itemName)
		}

		if i != 0 {
			s.WriteString(", ")
		} else {
			s.WriteString(".")
		}
	}
	return s.String()
}

func Song() string {

	var song strings.Builder

	for _, item := range items {
		song.WriteString(Verse(item.day))

		if item.day != 12 {
			song.WriteRune('\n')
		}
	}

	return song.String()
}